### PR TITLE
ENH Use class name instead of self

### PR DESCRIPTION
**Pull request by emteknetnz**
Issue https://github.com/silverstripe/developer-docs/issues/486

### CODE REVIEW

File: tests/php/ORM/Filters/EndsWithFilterTest.php [hunk 1]
```diff
@@ -267,7 +268,7 @@ public function testMatches(mixed $filterValue, mixed $matchValue, array $modifi
                 }
             }
 
-            EndsWithFilter::config()->set('default_case_sensitive', $caseSensitive);
+            SearchFilter::config()->set('default_case_sensitive', $caseSensitive);
```
Comment (emteknetnz) [inline tests/php/ORM/Filters/EndsWithFilterTest.php:271]: This was kind of working by accident before. The config in on SearchFilter, not EndsWithFilter

Ditto for the 3x other *FilterTest.php files

Comment (GuySartorelli) [reply tests/php/ORM/Filters/EndsWithFilterTest.php:271]: It wasn't by accident, that was intentional. You can change the config for each filter, fi you want to.

File: src/Control/Controller.php [hunk 2]
```diff
@@ -693,7 +693,7 @@ public static function normaliseTrailingSlash(string $url): string
         }
 
         // Normlise trailing slash
-        $shouldHaveTrailingSlash = self::config()->uninherited('add_trailing_slash');
+        $shouldHaveTrailingSlash = static::config()->uninherited('add_trailing_slash');
```
Comment (GuySartorelli) [inline src/Control/Controller.php]:
```suggestion
        $shouldHaveTrailingSlash = Controller::config()->uninherited('add_trailing_slash');
```
This really needs to affect the whole site uniformly. It doesn't make sense for this to change per subclass.

File: src/Core/Extensible.php [hunk 3]
```diff
@@ -124,7 +124,7 @@ protected function defineExtensionMethods()
         foreach ($extensions as $extensionClass => $extensionInstance) {
             foreach ($this->findMethodsFrom($extensionInstance) as $method) {
                 $this->addCallbackMethod($method, function ($inst, $args) use ($method, $extensionClass) {
-                    /** @var Extensible $inst */
+                    /** @var self $inst */
```
Comment (GuySartorelli) [inline src/Core/Extensible.php]: Why this change? Does the IDE not pick up on the methods if with `Extensible`, but it does with `self`?

File: src/ORM/Filters/SearchFilter.php [hunk 4]
```diff
@@ -474,7 +474,7 @@ protected function getCaseSensitive()
         } elseif (in_array('nocase', $modifiers ?? [])) {
             return false;
         } else {
-            $sensitive = self::config()->get('default_case_sensitive');
+            $sensitive = static::config()->get('default_case_sensitive');
```
Comment (GuySartorelli) [inline src/ORM/Filters/SearchFilter.php]:
```suggestion
            $sensitive = SearchFilter::config()->get('default_case_sensitive');
```
This config was added in https://github.com/silverstripe/silverstripe-framework/pull/10925 and was very much intended to be a single value across the full class hierarchy.

File: thirdparty/php-peg/Compiler.php
Comment (GuySartorelli) [inline thirdparty/php-peg/Compiler.php:1]: We probably shouldn't touch thirdparty - though if you think otherwise, I'm ultimately not too fussed.

**Review by GuySartorelli — CHANGES_REQUESTED** [review]

File: src/Control/Controller.php (hunk 2 above)
Comment (emteknetnz) [reply src/Control/Controller.php]: As this is a refactoring PR, I'm not changing any existing behaviour. This sort of change should happen in a separate PR, probably to CMS 6

File: src/Core/Extensible.php (hunk 3 above)
Comment (emteknetnz) [reply src/Core/Extensible.php]: Honestly I don't know, I've reverted it

File: thirdparty/php-peg/Compiler.php
Comment (emteknetnz) [reply thirdparty/php-peg/Compiler.php:1]: Yeah makes sense to exclude thirdparty, will revert

File: src/ORM/Filters/SearchFilter.php (hunk 4 above)
Comment (GuySartorelli) [reply src/ORM/Filters/SearchFilter.php]: This goes along with the change you already made - https://github.com/silverstripe/silverstripe-framework/pull/11270#discussion_r1627033946

The alternative is to swap the tests back to how they were, but one of those changes should be made.

File: src/Control/Controller.php (hunk 2 above)
Comment (GuySartorelli) [reply src/Control/Controller.php]: This seems like the same sort of change that you did agree to make in https://github.com/silverstripe/silverstripe-cms/pull/2955#discussion_r1642019070? I won't argue though - I can create a new issue for it. Won't be for CMS 6 though, we'd be fixing a bug and there's no API break.

Comment (emteknetnz) [reply src/Control/Controller.php]: OK, given that it would be fixed as a bug I guess makes sense to just just to `Controller::config()` now

File: src/ORM/Filters/SearchFilter.php (hunk 4 above)
Comment (emteknetnz) [reply src/ORM/Filters/SearchFilter.php]: Will change to SearchFilter::config()

**Review by GuySartorelli — APPROVED** [review]
LGTM
